rfcs/20230418-constant-tensor-cache-capacity-control/README.md

```diff
@@ -0,0 +1,266 @@
+consumption through this API. And users are suggested to set the workspace size
+as big as possible to achieve best performance. The default workspace size is
+set to device global memory size from v8.4 release.
+
+## Proposal
+
+### Dedicated cache for different engine kind
+In oneDNN Graph programming model, different graph can have different engine
+kind and simultaneously run on different engine, like cpu or gpu. Different kind
+of engine usually has different memory limitation. 
+
+So, we propose to use dedicated cache for different engine kind, and let users
+set different capacity for different engine kind separately.
+
+For multiple devices scenarios, like multi-GPU, whether to use dedicated cache
+for each device is still an open question, we will discuss it in later sections.
+
+### Unified cache for all backends
+We propose to implement an unified constant tensor cache in the oneDNN Graph
+interface layer. Each backend should use this interface to cache their all
+constant tensors into the unified cache. And the oneDNN Graph API users could
+control the capacity of the unified cache, no need to be aware of any backend
+specific things. This proposal is based on the following considerations:
+
+- In memory limited scenarios, we usually want to control the total memory
+  footprint, instead of the memory footprint of specific backend.
+- By the design of oneDNN Graph API, the API users don't need to be aware about
+  which backend is using. We don't like to break this design currently.
+- Backend selection is dynamic in mixed backend scenarios. Users can't
+  explicitly know which backend will run and how many partition will a backend
+  claims until graph partition stage done. So, it's hard to decide how many
+  memory a backend should use.
+
+### Capacity setting/getting frontend API
+Based on above considerations, we proposed to add the following two C++ API and
+corresponding C API to control the capacity of constant tensor cache:
+
+```cpp
+void set_constant_tensor_cache_capacity(dnnl::engine::kind eng_kind, int64_t capacity);
+int64_t get_constant_tensor_cache_capacity(dnnl::engine::kind eng_kind);
+```
+
+To limit the capacity for a specific engine kind users must call the setter API.
+The unit of set capacity API is megabytes (MB). When cached tensors total size
+reaches the capacity, new tensors won't be cached. To query current capacity for
+a specific engine kind user must call the getter API.
+
+The default capacity is kept to be zero on both CPU and GPU to keep the
+library default behavior same with before.
+
+In addition to programmable API there is an environment variable named
+`ONEDNN_GRAPH_CONSTANT_TENSOR_CACHE_CAPACITY` to control the capacity. It
+accepts values in form `engine_kind:size`. or
+`engine_kind1:size1;engine_kind2:size2`. Example:
+
+```shell
+# set cpu/gpu constant tensor cache capacity to 10 GB and 2 GB separately
+export ONEDNN_GRAPH_CONSTANT_TENSOR_CACHE_CAPACITY=cpu:10240;gpu:2048 
```

Comment:
```suggestion
export ONEDNN_GRAPH_CONSTANT_TENSOR_CACHE_CAPACITY=cpu:10240;gpu0:2048 ;gpu1:1024
```

Reply:
Actually, the current proposal doesn't support this scenario. We propose to use the same capacity for each device which means gpu0 and gpu1 will use the same capacity size. The reasons are:
1. It is relatively simple for API users to use. Currently, it may exhaust users if they need to set the capacity for each device because currently default behavior is zero capacity. 
2. For now, we don't have such requirements. And this can be an open question and we can support it once needed.(The open question section is removed from RFC)

I will add more descriptions to RFC. What do you think of it? I believe the below question about changing `engine_kind`-->`engine` should be the same question?